bot/exts/events/trivianight/_questions.py

```diff
@@ -0,0 +1,240 @@
+            question_embed.add_field(name=f"Answer {label}", value=answer, inline=False)
+
+        question_embed.set_footer(text="0 people have answered")
+        current_time = perf_counter()
+        for button in self.buttons:
+            button._time = current_time
+
+        time_limit = self.current_question.get("time", 10)
+
+        self.active_question = True
+
+        return question_embed, time_limit
+
+    def end_question(self) -> tuple[dict, Embed]:
+        """Returns the dictionaries from the corresponding buttons for those who got it correct."""
+        labels = self.current_labels
+        label = labels[self.current_question["answers"].index(self.current_question["correct"])]
+        return_dict = {name: (*info, info[0] == label) for name, info in self.users_picked.items()}
+        all_players = list(self.users_picked.items())
+
+        answer_embed = Embed(
+            title=f"The correct answer for Question {self.current_question['number']} was..",
+            description=self.current_question["correct"]
+        )
+
+        if len(all_players) != 0:
+            answers_chosen = {
+                answer_choice: len(
+                    tuple(filter(lambda x: x[0] == answer_choice, self.users_picked.values()))
+                ) / len(all_players)
+                for answer_choice in labels
+            }
+
+            for idx, (answer, percent) in enumerate(answers_chosen.items()):
+                # Setting the color of answer_embed to the % of people that got it correct via the mapping
+                if idx == 0:
+                    # Maps the % of people who got it right to a color, from a range of red to green
+                    percentage_to_color = [0xFC94A1, 0xFFCCCB, 0xCDFFCC, 0xB0F5AB, 0xB0F5AB]
+                    answer_embed.color = percentage_to_color[round(percent * 100) // 25]
+
+                # The `ord` function is used here to change the letter to its corresponding position
+                answer_embed.add_field(
+                    name=f"{percent * 100:.1f}% of players chose",
+                    value=self.current_question['answers'][ord(answer) - 65],
+                    inline=False
+                )
+
+        self.users_picked = {}
+
+        for button in self.buttons:
+            button.users_picked = self.users_picked
+
+        time_limit = self.current_question.get("time", 10)
+        question_points = self.current_question.get("points", 10)
+
+        self.active_question = False
+
+        return return_dict, answer_embed, time_limit, question_points
+
+
+class Questions:
+    """An interface to use from the TriviaNight cog for questions."""
+
+    def __init__(self, scoreboard: Scoreboard):
+        self.scoreboard = scoreboard
+        self.questions = []
+
+    def set_questions(self, questions: list) -> None:
+        """Setting `self.questions` dynamically via a function to set it."""
+        self.questions = questions
```

Comment:
Maybe we can just make a copy of the list (using `[:]`) and then pop/remove items from it rather than modifying it with the `visited` key?

Is there a reason you went for setting that key rather than something like this?

Reply:
Right, but instead of filtering that if we instead just make a copy as I suggested you can just iterate through all questions.

Reply:
That sounds fine, but how would we check if the question has been visited or not? Just compare it to a main list?

That main list would still be mutated though and it would be extra work to switch it though. Not sure if I'm following your suggestion tho

Reply:
Ohh, I see what you're saying. We have a list of all the questions which is just a dictionary and then pop the dictionaries out if they were already used?

Reply:
Yes - make a copy of the list from the deserialized JSON (otherwise you would modify the original list) and remove questions as they are picked (for example: `my_list.pop(random.randrange(len(my_list)))`.